### PR TITLE
Refactors the forensics component into a datum

### PR DESCRIPTION
**Pull request by SuperNovaa41**
## About The Pull Request

This was doing things component really shouldn't be doing, and now all
of its behaviour is contained onto a datum, as it should've been the
whole time

## Why It's Good For The CODEBASE

some things just really shouldn't be components, this was made back when DCS was first implemented and just thrown in because it was the new hot thing i guess, but datumized forensics makes far more sense, AND doesn't use GetComponent

TODO:

- [ ] More thorough testing to make sure nothing broke

## Changelog

:cl:
refactor: Turned the forensics component into a datum.
/:cl:

### CODE REVIEW

File: code/game/atoms.dm [hunk 1]
```diff
@@ -160,6 +160,8 @@
 
 	var/resistance_flags = NONE // INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | ON_FIRE | UNACIDABLE | ACID_PROOF
 
+	var/datum/forensics/forensics
```
Comment (Kylerace) [inline code/game/atoms.dm:163]: doc

Comment (SuperNovaa41) [reply code/game/atoms.dm:163]: whoops

File: code/modules/forensics/_forensics.dm [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	 * List of clothing fibres on this atom
+	 *
+	 * Formatting:
+	 * * fiber = fibre
+	 */
+	var/list/fibers
+
+/datum/forensics/New(atom/parent, fingerprints, hiddenprints, blood_DNA, fibers)
+	if(!isatom(parent))
+		return
```
Comment (MrMelbert) [inline code/modules/forensics/_forensics.dm:44]: Should probably stack trace, and `qdel(src)` if parent is ever not an atom

File: code/modules/forensics/_forensics.dm [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	LAZYINITLIST(src.hiddenprints)
+	for(var/hidden_print in hiddenprints) //We use an associative list, make sure we don't just merge a non-associative list into ours.
+		src.hiddenprints[hidden_print] = hiddenprints[hidden_print]
+	return TRUE
+
+/// Adds a single hiddenprint
+/datum/forensics/proc/add_hiddenprint(mob/suspect)
+	if(!isliving(suspect))
+		if(!iscameramob(suspect))
```
Comment (Kylerace) [inline code/modules/forensics/_forensics.dm:190]: the !isliving() can be replaced by isaicamera()

File: code/modules/forensics/_forensics.dm [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	var/hasgloves = ""
+	if(ishuman(suspect))
+		var/mob/living/carbon/human/human_suspect = suspect
+		if(human_suspect.gloves)
+			hasgloves = "(gloves)"
+	var/current_time = time_stamp()
+	if(!LAZYACCESS(hiddenprints, suspect.key))
+		LAZYSET(hiddenprints, suspect.key, "First: \[[current_time]\] \"[suspect.real_name]\"[hasgloves]. Ckey: [suspect.ckey]")
+	else
+		var/laststamppos = findtext(LAZYACCESS(hiddenprints, suspect.key), "\nLast: ")
```
Comment (Kylerace) [inline code/modules/forensics/_forensics.dm:208]: snake_case

File: code/modules/forensics/_forensics.dm [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		if(!iscameramob(suspect))
+			return
+		if(isaicamera(suspect))
+			var/mob/camera/ai_eye/ai_camera = suspect
+			if(!ai_camera.ai)
+				return
+			suspect = ai_camera.ai
+	if(!suspect.key)
+		return
+	var/hasgloves = ""
```
Comment (Kylerace) [inline code/modules/forensics/_forensics.dm:201]: snake_case

File: code/modules/forensics/_forensics.dm [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+	 *
+	 * Formatting:
+	 * * dna = bloodtype
+	 */
+	var/list/blood_DNA
+	/**
+	 * List of clothing fibres on this atom
+	 *
+	 * Formatting:
+	 * * fiber = fibre
```
Comment (Kylerace) [inline code/modules/forensics/_forensics.dm:38]: why are they spelled different.

File: code/modules/forensics/_forensics.dm [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		LAZYSET(fingerprints, full_print, full_print)
+	return TRUE
+
+/// Adds the given list into fibers
+/datum/forensics/proc/add_fiber_list(list/fibers)
+	if(!length(fibers))
+		return
+	LAZYINITLIST(src.fibers)
+	for(var/fibre in fibers) //We use an associative list, make sure we don't just merge a non-associative list into ours.
+		src.fibers[fibre] = fibre
```
Comment (Kylerace) [inline code/modules/forensics/_forensics.dm:139]: _ahem_ AHHHHHHHHHHHHHHHHHHH

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/modules/forensics/_forensics.dm (hunk 6 above)
Comment (SuperNovaa41) [reply code/modules/forensics/_forensics.dm:38]: sorry my canadian fingers keep trying to correct it

File: code/game/atoms.dm [hunk 8]
```diff
@@ -159,6 +159,8 @@
 	var/damage_deflection = 0
 
 	var/resistance_flags = NONE // INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | ON_FIRE | UNACIDABLE | ACID_PROOF
+	/// forensics datum, contains fingerprints, fibres, blood_dna and hiddenprints on this atom
+	var/datum/forensics/forensics
```
Comment (MrMelbert) [inline code/game/atoms.dm:163]: you need to handle this reference in atom `Destroy()`

i don't know off the top of my head if it should be `qdel`'d or just `null`ed, but you should probably check what the current behavior is (i think forensics gets qdel'd when its parent does, as it's a component)

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/examine.dm [hunk 9]
```diff
@@ -46,11 +46,10 @@
 		if(!(I.item_flags & ABSTRACT) && !(I.item_flags & EXAMINE_SKIP))
 			. += "[t_He] [t_is] holding [I.get_examine_string(user)] in [t_his] [get_held_index_name(get_held_index_of_item(I))]."
 
-	var/datum/component/forensics/FR = GetComponent(/datum/component/forensics)
 	//gloves
 	if(gloves && !(obscured & ITEM_SLOT_GLOVES) && !(gloves.item_flags & EXAMINE_SKIP))
 		. += "[t_He] [t_has] [gloves.get_examine_string(user)] on [t_his] hands."
-	else if(FR && length(FR.blood_DNA))
+	else if(forensics && length(forensics.blood_DNA))
```
Comment (MrMelbert) [inline code/modules/mob/living/carbon/human/examine.dm:52]: this should've been using the helper getters anyways
doesn't matter much now, but might as well enforce it for consistency(?)

```suggestion
	else if(return_blood_DNA())
```
or
```suggestion
	else if(blood_DNA_length())
```

File: code/modules/forensics/_forensics.dm [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	if (hiddenprints)
+		src.hiddenprints = LAZY_LISTS_OR(src.hiddenprints, hiddenprints)
+	if (blood_DNA)
+		src.blood_DNA = LAZY_LISTS_OR(src.blood_DNA, blood_DNA)
+	if (fibers)
+		src.fibers = LAZY_LISTS_OR(src.fibers, fibers)
+	check_blood()
+
+/datum/forensics/Destroy(force, ...)
+	UnregisterSignal(parent.resolve(), list(COMSIG_COMPONENT_CLEAN_ACT))
```
Comment (MrMelbert) [inline code/modules/forensics/_forensics.dm:70]: as parent's a weakref, you may need to check that this resolve is successful when it's done.
in other cases, this'll probably never come into play, but in `Destroy()` all bets are off. 
```
var/atom/real_parent = parent.resolve()
if(real_parent)
   ...
```

Comment (SuperNovaa41) [reply code/modules/forensics/_forensics.dm:70]: gotcha

File: code/__DEFINES/blood.dm [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /// Checks if an object is covered in blood
-#define HAS_BLOOD_DNA(thing) (length(thing.GetComponent(/datum/component/forensics)?.blood_DNA))
+#define HAS_BLOOD_DNA(thing) (length(thing.forensics?.blood_DNA))
```
Comment (MrMelbert) [inline code/__DEFINES/blood.dm:2]: Now that I think about it, why does this macro exist when the helper procs (`blood_DNA_length` and `return_blood_DNA`) exist?
or why do the helper procs exist when they could be macros?
🤔

**Pull request merged**